src/mumble/PipeWire.cpp

```diff
@@ -0,0 +1,427 @@
+	pws->m_lock.unlock();
+
+	if (pws->m_ok) {
+		inputRegistrar  = std::make_unique< PipeWireInputRegistrar >();
+		outputRegistrar = std::make_unique< PipeWireOutputRegistrar >();
+	} else {
+		pws.reset();
+	}
+}
+
+void PipeWireInit::destroy() {
+	inputRegistrar.reset();
+	outputRegistrar.reset();
+	pws.reset();
+}
+
+static PipeWireInit pwi;
+
+PipeWireSystem::PipeWireSystem() : m_ok(false), m_users(0) {
+	const QStringList names{ "libpipewire.so", "libpipewire-0.3.so" };
+
+	for (const auto &name : names) {
+		m_lib.setFileName(name);
+		if (m_lib.load()) {
+			break;
+		}
+	}
+
+	if (!m_lib.isLoaded()) {
+		return;
+	}
+
+	RESOLVE(pw_get_library_version);
+	RESOLVE(pw_init);
+	RESOLVE(pw_deinit);
+	RESOLVE(pw_loop_new);
+	RESOLVE(pw_loop_destroy);
+	RESOLVE(pw_thread_loop_new_full);
+	RESOLVE(pw_thread_loop_destroy);
+	RESOLVE(pw_thread_loop_start);
+	RESOLVE(pw_thread_loop_stop);
+	RESOLVE(pw_properties_new);
+	RESOLVE(pw_stream_new_simple);
+	RESOLVE(pw_stream_destroy);
+	RESOLVE(pw_stream_connect);
+	RESOLVE(pw_stream_dequeue_buffer);
+	RESOLVE(pw_stream_queue_buffer);
+
+	qInfo("PipeWire %s from %s", pw_get_library_version(), qPrintable(m_lib.fileName()));
+
+	pw_init(nullptr, nullptr);
+
+	m_ok = true;
+}
+
+PipeWireSystem::~PipeWireSystem() {
+	if (m_ok) {
+		pw_deinit();
+	}
+}
+
+PipeWireEngine::PipeWireEngine(const char *category, void *param, const std::function< void(void *param) > callback)
+	: m_ok(false), m_loop(nullptr), m_stream(nullptr), m_thread(nullptr) {
+	if (!pws->isOk()) {
+		return;
+	}
+
+	m_loop = pws->pw_loop_new(nullptr);
+	if (!m_loop) {
+		return;
+	}
+
+	pw_properties *props =
+		pws->pw_properties_new(PW_KEY_MEDIA_NAME, "Mumble", PW_KEY_MEDIA_CATEGORY, category, PW_KEY_MEDIA_TYPE, "Audio",
+							   PW_KEY_MEDIA_ROLE, "Communication", nullptr);
+
+	m_events          = std::make_unique< pw_stream_events >();
+	m_events->version = PW_VERSION_STREAM_EVENTS;
+	m_events->process = *callback.target< void (*)(void *) >();
+
+	m_stream = pws->pw_stream_new_simple(m_loop, category, props, m_events.get(), param);
+	if (!m_stream) {
+		return;
+	}
+
+	m_thread = pws->pw_thread_loop_new_full(m_loop, category, nullptr);
+	if (m_thread) {
+		m_ok = true;
+	}
+}
+
+PipeWireEngine::~PipeWireEngine() {
+	if (!pws->isOk()) {
+		return;
+	}
+
+	if (m_thread) {
+		pws->pw_thread_loop_destroy(m_thread);
+	}
+
+	if (m_stream) {
+		pws->pw_stream_destroy(m_stream);
+	}
+
+	if (m_loop) {
+		pws->pw_loop_destroy(m_loop);
+	}
+}
+
+bool PipeWireEngine::connect(const uint8_t direction, const uint32_t *channels, const uint8_t nChannels) {
+	uint8_t buffer[1024];
+
+	spa_pod_builder builder{};
+	builder.data = buffer;
+	builder.size = sizeof(buffer);
+
+	spa_audio_info_raw info{};
+	info.channels = nChannels;
+	info.format   = SPA_AUDIO_FORMAT_F32;
+	info.rate     = SAMPLE_RATE;
+
+	if (nChannels < 2) {
+		info.position[0] = SPA_AUDIO_CHANNEL_MONO;
+	} else {
+		for (uint8_t i = 0; i < nChannels; ++i) {
```

Comment:
Just FYI: fixed-width integers can actually end up being slower than using plain `(unsigned) int` due to processors having specific register sizes.

Reply:
Right, I would expect the compiler to take care of that though.

The only reason I opted for `uint8_t` is to avoid allocating unnecessary extra bytes, as the standard audio channels are 18: https://en.wikipedia.org/wiki/Surround_sound#Standard_speaker_channels

https://github.com/mumble-voip/mumble/blob/003b6c6bb2e993a51fa7f46429fc646bacefe471/src/mumble/AudioOutput.h#L23-L42

Reply:
> Right, I would expect the compiler to take care of that though.

I don't think it is allowed to do so. If you want that, you'd have to use `std::int_fast8_t` instead

Reply:
Or `uint_fast8_t`: https://en.cppreference.com/w/c/types/integer

However, I don't think it's worth it, the function is only called once after all.

Reply:
I don't think so either. But what I was trying to get at: Using a plain `int` in these cases would probably be faster and would then pertain the special meaning of where fixed-width integers are used :)

Reply:
Sure! :)